src/v/transform/transform_manager.cc

```diff
@@ -0,0 +1,386 @@
+            entry_t,
+            boost::multi_index::
+              const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+            boost::multi_index::
+              const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>>,
+          boost::multi_index::composite_key_hash<
+            std::hash<model::transform_id>,
+            std::hash<model::ntp>>>,
+        // indexed by ntp
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<ntp_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>,
+          std::hash<model::ntp>,
+          std::equal_to<>>,
+        // indexed by id
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<id_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+          std::hash<model::transform_id>,
+          std::equal_to<>>>>;
+
+public:
+    const entry_t& insert(
+      std::unique_ptr<processor> processor,
+      ss::lw_shared_ptr<transform::probe> probe) {
+        auto [it, inserted] = _underlying.emplace(
+          std::move(processor), std::move(probe));
+        vassert(inserted, "invalid transform processor management");
+        return *it;
+    }
+
+    ss::lw_shared_ptr<probe> get_or_create_probe(
+      model::transform_id id, const model::transform_metadata& meta) {
+        auto& id_index = _underlying.get<id_index_t>();
+        auto id_it = id_index.find(id);
+        if (id_it == id_index.end()) {
+            auto probe = ss::make_lw_shared<transform::probe>();
+            probe->setup_metrics(meta.name);
+            return probe;
+        }
+        return id_it->probe();
+    }
+
+    bool contains(model::transform_id id, const model::ntp& ntp) {
+        auto& by_key = _underlying.get<key_index_t>();
+        return by_key.contains(std::make_tuple(id, ntp));
+    }
+
+    const entry_t* get_or_null(model::transform_id id, const model::ntp& ntp) {
+        auto& by_key = _underlying.get<key_index_t>();
+        auto it = by_key.find(std::make_tuple(id, ntp));
+        if (it == by_key.end()) {
+            return nullptr;
+        }
+        return std::addressof(*it);
+    }
+
+    ss::future<> clear() {
+        co_await ss::parallel_for_each(
+          _underlying.begin(),
+          _underlying.end(),
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [](auto& e) { return e.processor()->stop(); });
+        _underlying.clear();
+    }
+
+    ss::future<> erase_by_id(model::transform_id id) {
+        auto& by_id = _underlying.get<id_index_t>();
+        auto range = by_id.equal_range(id);
+        co_await ss::parallel_for_each(
+          range.first,
+          range.second,
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [](auto& e) { return e.processor()->stop(); });
```

Comment:
for all of these futurized methods that operate on the underlying container, what protection is there from concurrent modification of the container?

Reply:
That's the work queue - it ensures that we only ever do things sequentially and there are no concurrent races to worry about. I do see that there is a ss::parallel_for_each in side the queue. I'll remove that for now because it will simplify things and I'm trying not to prematurely optimize.

Reply:
I missed that usage, thanks for pointing it out.

IMO this could serve as a model for building more reliable control plane services. We had noodled doing something similar before, but always ended up building the harder to analyze concurrent versions even when performance wasn't critical.

Reply:
> I missed that usage, thanks for pointing it out.
> IMO this could serve as a model for building more reliable control plane services. We had noodled doing something similar before, but always ended up building the harder to analyze concurrent versions even when performance wasn't critical.


Totally. I think this pattern is good for things that just need to be simple. Hopefully over time the queue can have more abilities (like running IO off the queue and having a continuation run on the queue with the result) and this can be a good default. I've already seen a few people miss a lock or forget to close a gate, so I think there can be a lot of value in having a component that handles those bits for you.

This essentially moves services within a core to forcing message passing (much like how seastar forces this between cores), which ends up simplifying a lot of concurrency issues that just get tricky to reason about when locks, and the like get involved.